keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,733 @@
+     that the cloud returned at least one IP of the intended family,
+     this will result in that becoming the Primary Node IP.
+
+  3. (If the user passes no `--node-ip` the list of addresses from the
+     cloud provider is used unchanged.)
+
+On bare metal, if an explicit `--node-ip` has been passed, then
+kubelet will create a single `InternalIP` address with that IP.
+Otherwise it will try:
+
+  1. Parsing the node's hostname as an IP address
+  2. Looking up the node's node name (not hostname?!) in DNS
+  3. Finding an IP address on the node's default NIC
+
+If `--node-ip ::` was passed, then IPv6 DNS/interface addresses will
+be preferred; otherwise IPv4 addresses will be preferred.
+
+Although these rules can end up choosing an IPv6 Primary Node IP when
+no `--node-ip` was passed in some situations, this will not actually
+result in a fully-functioning node, since kubelet will set up
+IPv4-only iptables rules if no `--node-ip` was passed.
+
+#### IPv6 Addresses in `Node.Status.Addresses`
+
+Most cloud providers do not currently include IPv6 addresses in
+`Node.Status.Addresses`. (See the "[Cloud Provider
+Survey](#cloud-provider-survey)" at the end.) To some extent this is
+just because they haven't been updated for IPv6/Dual-Stack support,
+but this is also because people are worried about confusing existing
+clients by adding IPv6 addresses to a field that historically only
+held IPv4 addresses. (Something we have also worried about in other
+places. eg, Endpoints.)
+
+Some cloud providers (notably Azure) are already returning IPv6 node
+IPs when they exist, without apparent problems. OTOH, the fact that
+there haven't been problems with clusters on Azure (a relatively new
+platform) may not mean that there wouldn't be problems with clusters
+on AWS if the AWS CloudProvider suddenly started returning IPv6
+addresses.
+
+On the other other hand, CloudProviders will not be able to return
+IPv6 addresses unless the Node actually _has_ IPv6 addresses, and on
+most platforms it probably won't unless the user explicitly chose to
+have them. So in that case, making CloudProviders suddenly start
+returning IPv6 addresses when they exist still won't cause problems,
+even in clusters where there is old software that only handles IPv4
+addresses in `Node.Status.Addresses`, since the clusters running the
+old software presumably don't have any IPv6 node addresses anyway.
+
+If we are concerned about confusing old software, there are three
+approaches we could take to fix it:
+
+  1. Add one or more new `v1.NodeAddressType` values to use for IPv6
+     addresses. eg, `InternalIPv6` and `ExternalIPv6`. One problem
+     with this is that there is code scattered across many components
+     (and, eg, network plugins) that assumes that all Nodes have
+     either an `InternalIP` address or an `ExternalIP` address, so
+     using different types for IPv6 addresses would likely break
+     single-stack IPv6 clusters.
```

Comment:
In the case of single-stack IPv6, I think it would be acceptable to continue using InternalIP and ExternalIP for those addresses.